File: debris.py
"""SATCAT shell crowding screen for LEO collision-avoidance budgeting.

This is a population screen, not operational collision probability (Pc). SATCAT
counts objects in altitude shells; it does not propagate objects in 3D.
"""
from __future__ import annotations

import json
import time
from pathlib import Path
from typing import Any, Iterable
from urllib.request import urlopen

SATCAT_URL = "https://celestrak.org/satcat/records.php?FORMAT=JSON&ONORBIT=1"
CACHE_PATH = Path(__file__).resolve().parent / ".cache" / "satcat_onorbit.json"
CACHE_TTL_S = 6 * 60 * 60

# Labeled fallback: deliberately small and synthetic, never presented as SATCAT.
SYNTHETIC_SATCAT = [
    *({"APOGEE": 545, "PERIGEE": 545, "OBJECT_TYPE": "PAY", "OBJECT_NAME": "synthetic payload"} for _ in range(18)),
    *({"APOGEE": 540, "PERIGEE": 540, "OBJECT_TYPE": "DEB", "OBJECT_NAME": "synthetic debris"} for _ in range(14)),
    *({"APOGEE": 400, "PERIGEE": 400, "OBJECT_TYPE": "R/B", "OBJECT_NAME": "synthetic rocket body"} for _ in range(5)),
    *({"APOGEE": 400, "PERIGEE": 400, "OBJECT_TYPE": "PAY", "OBJECT_NAME": "synthetic reference payload"} for _ in range(3)),
]


def _valid_records(value: Any) -> list[dict[str, Any]]:
    if not isinstance(value, list):
        raise ValueError("SATCAT response was not a JSON array")
    return [row for row in value if isinstance(row, dict)]


def load_satcat(force_refresh: bool = False) -> list[dict[str, Any]]:
    """Load on-orbit SATCAT records, caching for about six hours.

    Network and cache errors intentionally return a labeled synthetic LEO mix so
    a simulation remains usable when CelesTrak is unavailable.
    """
    try:
        if not force_refresh and CACHE_PATH.exists() and time.time() - CACHE_PATH.stat().st_mtime < CACHE_TTL_S:
            return _valid_records(json.loads(CACHE_PATH.read_text(encoding="utf-8")))
        with urlopen(SATCAT_URL, timeout=15) as response:
            records = _valid_records(json.loads(response.read().decode("utf-8")))
        CACHE_PATH.parent.mkdir(parents=True, exist_ok=True)
        CACHE_PATH.write_text(json.dumps(records), encoding="utf-8")
        return records
    except Exception:
        return [dict(row) for row in SYNTHETIC_SATCAT]


def cam_scale_from_counts(shell_total: int, reference_total: int) -> float:
    return max(0.4, min(2.5, shell_total / max(reference_total, 1)))


def _altitude(row: dict[str, Any]) -> tuple[float, float] | None:
    try:
        apogee, perigee = float(row["APOGEE"]), float(row["PERIGEE"])
    except (KeyError, TypeError, ValueError):
        return None
    if min(apogee, perigee) < 80 or max(apogee, perigee) > 2000:
        return None
    return apogee, perigee


def _earth(row: dict[str, Any]) -> bool:
    return row.get("ORBIT_CENTER") in (None, "", "EA")


def _counts(rows: Iterable[dict[str, Any]], altitude_km: float, half_width_km: float) -> dict[str, int]:
    counts = {"payload": 0, "rocket_body": 0, "debris": 0, "unknown": 0}
    for row in rows:
        if not _earth(row):
            continue
        bounds = _altitude(row)
        if bounds is None or not (altitude_km - half_width_km <= sum(bounds) / 2 <= altitude_km + half_width_km):
            continue
        kind = {"PAY": "payload", "R/B": "rocket_body", "DEB": "debris"}.get(row.get("OBJECT_TYPE"), "unknown")
        counts[kind] += 1
    counts["total"] = sum(counts.values())
    return counts


def shell_census(satcat: Iterable[dict[str, Any]], altitude_km: float, half_width_km: float = 30, reference_alt_km: float = 400) -> dict[str, Any]:
    rows = list(satcat)
    counts = _counts(rows, altitude_km, half_width_km)
    reference = _counts(rows, reference_alt_km, half_width_km)
    scale = cam_scale_from_counts(counts["total"], reference["total"])
    total = counts["total"]
    obstacle = "quiet" if scale < 0.85 else "nominal" if scale <= 1.25 else "crowded" if scale <= 1.8 else "severe"
    source = "CelesTrak SATCAT" if rows and not rows == SYNTHETIC_SATCAT else "synthetic fallback"
    note = "Screening shell count; not collision probability (Pc). Close approaches use a 5 km screening volume, not NASA-grade Pc."
    return {"altitude": altitude_km, "shell_low": altitude_km - half_width_km, "shell_high": altitude_km + half_width_km, "counts": counts, "reference_counts": reference, "cam_scale": scale, "obstacle": obstacle, "source": source, "note": note}

File: web/src/sim/orbital/debris.ts
export type SatcatObjectType = "PAY" | "R/B" | "DEB" | "UNK";

export interface SatcatRecord {
  APOGEE?: number | string | null;
  PERIGEE?: number | string | null;
  OBJECT_TYPE?: SatcatObjectType | string | null;
  ORBIT_CENTER?: string | null;
  OBJECT_NAME?: string | null;
  NORAD_CAT_ID?: number | string | null;
  INCLINATION?: number | string | null;
}

export interface ShellCounts {
  payload: number;
  rocket_body: number;
  debris: number;
  unknown: number;
  total: number;
}

export interface SatcatCensus {
  altitude: number;
  shell_low: number;
  shell_high: number;
  counts: ShellCounts;
  reference_counts: ShellCounts;
  cam_scale: number;
  obstacle: "quiet" | "nominal" | "crowded" | "severe";
  source: string;
  note: string;
}

export const SATCAT_URL =
  "https://celestrak.org/satcat/records.php?FORMAT=JSON&ONORBIT=1";

export const SYNTHETIC_SATCAT: SatcatRecord[] = [
  ...Array.from({ length: 18 }, () => ({ APOGEE: 545, PERIGEE: 545, OBJECT_TYPE: "PAY" })),
  ...Array.from({ length: 14 }, () => ({ APOGEE: 540, PERIGEE: 540, OBJECT_TYPE: "DEB" })),
  ...Array.from({ length: 5 }, () => ({ APOGEE: 400, PERIGEE: 400, OBJECT_TYPE: "R/B" })),
  ...Array.from({ length: 3 }, () => ({ APOGEE: 400, PERIGEE: 400, OBJECT_TYPE: "PAY" })),
];

const CACHE_KEY = "apogee.satcat.onorbit";
const CACHE_TTL_MS = 6 * 60 * 60 * 1000;

export function camScaleFromCounts(shellTotal: number, referenceTotal: number): number {
  return Math.max(0.4, Math.min(2.5, shellTotal / Math.max(referenceTotal, 1)));
}

function numeric(value: number | string | null | undefined): number | null {
  const parsed = typeof value === "number" ? value : Number(value);
  return Number.isFinite(parsed) ? parsed : null;
}

function meanAltitude(row: SatcatRecord): number | null {
  const apogee = numeric(row.APOGEE);
  const perigee = numeric(row.PERIGEE);
  if (apogee === null || perigee === null || Math.min(apogee, perigee) < 80 || Math.max(apogee, perigee) > 2000) return null;
  return (apogee + perigee) / 2;
}

function counts(rows: SatcatRecord[], altitude: number, halfWidth: number): ShellCounts {
  const result: ShellCounts = { payload: 0, rocket_body: 0, debris: 0, unknown: 0, total: 0 };
  for (const row of rows) {
    if (row.ORBIT_CENTER && row.ORBIT_CENTER !== "EA") continue;
    const mean = meanAltitude(row);
    if (mean === null || Math.abs(mean - altitude) > halfWidth) continue;
    const key = row.OBJECT_TYPE === "PAY" ? "payload" : row.OBJECT_TYPE === "R/B" ? "rocket_body" : row.OBJECT_TYPE === "DEB" ? "debris" : "unknown";
    result[key] += 1;
    result.total += 1;
  }
  return result;
}

export function shellCensus(rows: SatcatRecord[], altitude: number, halfWidth = 30, referenceAltitude = 400, source = "synthetic fallback"): SatcatCensus {
  const shell = counts(rows, altitude, halfWidth);
  const reference = counts(rows, referenceAltitude, halfWidth);
  const camScale = camScaleFromCounts(shell.total, reference.total);
  const obstacle = camScale < 0.85 ? "quiet" : camScale <= 1.25 ? "nominal" : camScale <= 1.8 ? "crowded" : "severe";
  return {
    altitude,
    shell_low: altitude - halfWidth,
    shell_high: altitude + halfWidth,
    counts: shell,
    reference_counts: reference,
    cam_scale: camScale,
    obstacle,
    source,
    note: "Screening shell count; not collision probability (Pc). Close approaches use a 5 km screening volume, not NASA-grade Pc.",
  };
}

export async function loadSatcat(forceRefresh = false): Promise<SatcatRecord[]> {
  try {
    if (!forceRefresh) {
      const cached = localStorage.getItem(CACHE_KEY);
      if (cached) {
        const parsed = JSON.parse(cached) as { saved: number; rows: SatcatRecord[] };
        if (Date.now() - parsed.saved < CACHE_TTL_MS) return parsed.rows;
      }
    }
    const response = await fetch(SATCAT_URL);
    if (!response.ok) throw new Error(`SATCAT HTTP ${response.status}`);
    const rows = (await response.json()) as SatcatRecord[];
    if (!Array.isArray(rows)) throw new Error("SATCAT response was not an array");
    localStorage.setItem(CACHE_KEY, JSON.stringify({ saved: Date.now(), rows }));
    return rows;
  } catch {
    return SYNTHETIC_SATCAT.map((row) => ({ ...row }));
  }
}
